### DISCUSSION TITLE
Support custom desktop app icon to replace the default Python logo

### DISCUSSION THREAD

**Discussion by hhstore** [zauberzeug/nicegui#1005] [Ideas / Feature Requests]
### Question

- ✅️ I try to develop a `desktop app` based on nicegui.
- ✅️ Is there a way to change the `default app icon(python logo)`? which is ugly.
- ✅️ Like Tauri: https://tauri.app/v1/guides/features/icons/

<img width="46" alt="image" src="https://github.com/zauberzeug/nicegui/assets/3252130/db34c6c6-a81c-4f3d-93bc-2bb9c6d5f588">

**Comment by rodja** [comment]
I agree, the default icon is ugly and NiceGUI deserves better. I would really like to wait on PyWebview to support this because I fear quite some platform specifics which we then must maintain. But If someone comes up with a good proposal we are happy to discuss it.

**Reply by AriBermeki** [reply]
@rodja  
This behavior is not related to pywebview but to the Python launcher itself.
When you run a Python script directly with the interpreter, the process in the taskbar will always show the default Python icon, and this process icon cannot be changed at runtime. Only the window icon can be customized through libraries like pywebview, tkinter, or PyQt.

If you want a custom taskbar icon, you need to build your script into a standalone executable (e.g., with PyInstaller or Nuitka) and embed your own icon there.
So this limitation comes from Python/the interpreter, not from pywebview.

**Reply by AriBermeki** [reply]
@rodja 
@hhstore 

 here is [a detailed explanation on Stack Overflow](https://stackoverflow.com/questions/14900510/changing-the-application-and-taskbar-icon-python-tkinter) that confirms my statement.

**Reply by rodja** [reply]
Interesting. Would it be possible to integrate an option into [nicegui-pack](https://nicegui.io/documentation/section_configuration_deployment#package_for_installation)? Who is up for a pull request?

**Reply by AriBermeki** [reply]
# Changing Icon Function at Runtime + Packaging Details for NiceGUI-Pack

Icons cannot be changed **at runtime**.  
If we really want to support changing the icon during execution, this would need to be discussed on the **CPython level** – right now, the Python launcher blocks this.  
(In other languages like Rust, this restriction is not as strict.)  

---

## Packaging Question
I could implement your request, but I’d need a few more details first:  
- Which **dependencies** should `NiceGUI-Pack` include?  
- Which **files** need to be bundled in the package?  

### My preference: Nuitka over PyInstaller
For packaging, I personally prefer **Nuitka** instead of **PyInstaller**.  
- **Why Nuitka is better**: It compiles Python code to C/C++ and then to a binary, which usually gives better performance, smaller runtime overhead, and improved security compared to PyInstaller (which mostly just bundles the interpreter and scripts).  
- **Downside**: Nuitka takes noticeably more time to build compared to PyInstaller, which is much faster. So while PyInstaller is more convenient for quick builds, Nuitka produces a more optimized final package.  

---

## Status Update
Just for context: this may take some time.  
I’m based in Germany and the second exam phase at university in September is quite stressful at the moment.  

Regarding **#3529**, we’ve already made some progress:  
- Changed the framework name to PyOrion (since `PyFrame` already exists in the PyPI namespace).  
- Communication now runs **directly from Python to Rust via IPC**:  
  - Windows → Named Pipe  
  - Linux/macOS/Android/iOS → UDS protocol  

---

## Question for you
Do you think these changes could solve your problem?  

There’s also an example of how to integrate this into NiceGUI here:  
👉 [examples/nicegui/main.py](https://github.com/YellowSiC/pyorion/blob/main/examples/nicegui/main.py)  

---

Please build it locally for now. We’re clarifying the licensing before publishing on PyPI—see the README.md

---

Looking forward to your feedback! 🙂

**Comment by hhstore** [comment]
## Nuitka

- https://github.com/Nuitka/Nuitka#icons


```ruby

# These create binaries with icons on Windows
python -m nuitka --onefile --windows-icon-from-ico=your-icon.png program.py
python -m nuitka --onefile --windows-icon-from-ico=your-icon.ico program.py
python -m nuitka --onefile --windows-icon-template-exe=your-icon.ico program.py

# These create application bundles with icons on macOS
python -m nuitka --macos-create-app-bundle --macos-app-icon=your-icon.png program.py
python -m nuitka --macos-create-app-bundle --macos-app-icon=your-icon.icns program.py
```
